### ISSUE TITLE
Use integrated Windows authentication when connecting to AWS via Windows proxy

### ISSUE THREAD

**Issue by kryoflux** [aws/aws-sdk-java#1061]
In order to avoid having to store/maintain a username and password while connecting to AWS via an authenticating Windows proxy (a requirement of many an IT security department I'm sure, but at least here), the AWS SDK could use the features of the Apache HTTP client to do Windows "integrated authentication" / NTLM.

One light-weight way to do this would be to provide the ability for an SDK user to customise the Apache HttpClientBuilder (as created in the SDKs ApacheHttpClientFactory class), by being able to specify a CredentialsProvider and DefaultAuthSchemeRegistry given the otherwise extra optional dependencies Windows authentication would need (httpclient-win, JNA).

Is the SDK use of the Apache HTTP client classes meant to be shielded off from SDK users?

Does this feature (either supporting directly, or via permitted customisation) seem like something that would reasonably fit?

**Comment by kiiadi** [comment]
@kryoflux we'd rather not expose the internals of the http layer directly to *most* customers. We want the experience to be as simple as possible and also maintain the flexibility to move to another HTTP client implementation in the future if we choose to.

That said, in a future major-version bump, we're looking to completely decouple the SDK from any specific HTTP client implementation. We'll do this behind a set of SDK-defined interfaces. We'll still provide a default implementation of course (most likely the same Apache HTTP client we use today) but this will give customers the flexibility to implement their own HTTP layer (either by extending one of ours, or rolling their own). This is where we see low-level specific implementations like your integrated authentication suggestion going.

TL;DR - we want to keep the default simple, but build hooks to allow 'power-users' to customize to suit their requirements.

**Comment by kryoflux** [comment]
@kiiadi Thanks for the info. Separating off the HTTP client does seem like a good strategy.

I do notice while poking about that, if I supply a ClientConfiguration to the AmazonS3Client, there is a getApacheHttpClientConfig() method that currently supports setting a org.apache.http.conn.socket.ConnectionSocketFactory. So I suppose in that sense it already exposes it a bit, with a well named (Apache-specific, here-be-dragons) getter. :)

It's very tempting to send a pull request to expand this class to allow also specifying a CredentialsProvider and AuthSchemeRegistry as I'd rather not have to maintain my own AWS SDK fork... Is it likely this would be accepted given this existing Apache-specific functionality will need to be carved off in the mentioned major version bump anyway?

**Comment by kiiadi** [comment]
@kryoflux I don't have a huge problem with it as long as it's backwards compatible and has adequate testing - but you might be better off waiting for the newest MVP of the SDK! We're looking to have a developer preview out sometime soon. *(note: deliberately avoiding giving a specific date - it'll be ready when it's ready!)*

**Comment by debora-ito** [comment]
Hi @kryoflux, the SDK team has reviewed the feature request list for V1, and since they're concentrating efforts on V2 new features they decided to not implement this one in V1.

It's still being considered for V2. I'll go ahead and close this.

Please feel free to reach out if you have further questions.
